accelerator/common_core.c

```diff
@@ -284,68 +284,88 @@ status_t ta_find_transactions_by_tag(const iota_client_service_t* const service,
+                                     const ta_find_transaction_objects_req_t* const req, transaction_array_t* res) {
   status_t ret = SC_OK;
-  flex_trit_t tx_trits[FLEX_TRIT_SIZE_8019];
-  flex_trit_t hash_trits[NUM_TRITS_HASH];
-  flex_trits_from_trytes(hash_trits, NUM_TRITS_HASH, (const tryte_t*)req, NUM_TRYTES_HASH, NUM_TRYTES_HASH);
-  char cache_value[FLEX_TRIT_SIZE_8019] = {0};
-
-  // get raw transaction data of transaction hashes
-  get_trytes_req_t* get_trytes_req = get_trytes_req_new();
-  get_trytes_res_t* get_trytes_res = get_trytes_res_new();
-  if (get_trytes_req == NULL || get_trytes_res == NULL) {
-    ret = SC_CCLIENT_OOM;
-    goto done;
+  flex_trit_t tx_trits[NUM_TRITS_SERIALIZED_TRANSACTION];
+  iota_transaction_t* temp = NULL;
+  get_trytes_req_t* req_get_trytes = get_trytes_req_new();
+  transaction_array_t* uncached_txn_array = transaction_array_new();
+  if (req == NULL || res == NULL || req_get_trytes == NULL || uncached_txn_array == NULL) {
+    return SC_TA_NULL;
   }
+  char txn_hash[NUM_TRYTES_HASH + 1] = {0};
+  char cache_value[NUM_TRYTES_SERIALIZED_TRANSACTION + 1] = {0};
+  txn_hash[NUM_TRYTES_HASH] = '\0';
+  cache_value[NUM_TRYTES_SERIALIZED_TRANSACTION] = '\0';
 
-  // get in cache first, then get from full node if no result in cache
-  ret = cache_get(req, cache_value);
-  if (ret == SC_OK) {
-    flex_trits_from_trytes(tx_trits, NUM_TRITS_SERIALIZED_TRANSACTION, (const tryte_t*)cache_value,
-                           NUM_TRYTES_SERIALIZED_TRANSACTION, NUM_TRYTES_SERIALIZED_TRANSACTION);
-  } else {
-    ret = hash243_queue_push(&get_trytes_req->hashes, hash_trits);
-    if (ret) {
-      ret = SC_CCLIENT_HASH;
-      goto done;
+  // append transaction object which is already cached to transaction_array_t
+  // if not, append uncached to request object of `iota_client_find_transaction_objectss`
+  hash243_queue_entry_t* q_iter = NULL;
+  CDL_FOREACH(req->hashes, q_iter) {
+    flex_trits_to_trytes((tryte_t*)txn_hash, NUM_TRYTES_HASH, q_iter->hash, NUM_TRITS_HASH, NUM_TRITS_HASH);
+
+    ret = cache_get(txn_hash, cache_value);
+    if (ret == SC_OK) {
+      flex_trits_from_trytes(tx_trits, NUM_TRITS_SERIALIZED_TRANSACTION, (const tryte_t*)cache_value,
+                             NUM_TRYTES_SERIALIZED_TRANSACTION, NUM_TRYTES_SERIALIZED_TRANSACTION);
+
+      // deserialize raw data to transaction object
+      temp = transaction_deserialize(tx_trits, true);
+      if (temp == NULL) {
+        ret = SC_CCLIENT_OOM;
+        goto done;
+      }
+
+      transaction_array_push_back(res, temp);
+      transaction_free(temp);
+    } else {
+      if (hash243_queue_push(&req_get_trytes->hashes, q_iter->hash) != RC_OK) {
+        ret = SC_CCLIENT_OOM;
+        goto done;
+      }
+      ret = SC_OK;
     }
 
-    ret = iota_client_get_trytes(service, get_trytes_req, get_trytes_res);
-    if (ret) {
+    // reset the string `cache_value`
+    cache_value[0] = '\0';
```

Comment:
This operation could be moved to line 322, only reset `cache_value` when getting value from cache.